README.md
# pkce

[![Go Reference](https://pkg.go.dev/badge/github.com/matthewhartstonge/pkce.svg)](https://pkg.go.dev/github.com/matthewhartstonge/pkce)
[![Go Report Card](https://goreportcard.com/badge/github.com/matthewhartstonge/pkce)](https://goreportcard.com/report/github.com/matthewhartstonge/pkce)
[![Build](https://github.com/matthewhartstonge/pkce/actions/workflows/build.yml/badge.svg?branch=main)](https://github.com/matthewhartstonge/pkce/actions/workflows/build.yml)

`pkce` implements the client side of RFC 7636 "Proof Key for Code Exchange by OAuth Public Clients" (PKCE) to enable the
generation of cryptographically secure and specification compliant code verifiers and code challenges. With :sparkles:
no external dependencies :sparkles:.

## Getting Started

`pkce` makes use of go mod, you can install it by using go get:

```shell
go get github.com/matthewhartstonge/pkce
```

## Examples

### Structs

For those that like abstractions, feel free to enjoy "safety:tm:":

```go
package main

import (
	"fmt"

	"github.com/matthewhartstonge/pkce"
)

func main() {
	// Generate a secure proof key! 
	// Make sure yo do this each time you want a new proof key - it's stateful.
	key, err := pkce.New()
	if err != nil {
		panic(err)
	}

	fmt.Println("my generated code verifier is:", key.CodeVerifier())
	fmt.Println("my generated plain code challenge is:", key.CodeChallenge())

	// Finally - on the server-side side, we can verify the received code 
	// verifier:
	receivedCodeVerifier := "#yolo-cant-verify-me-mr-mcbaggins"
	isValid := key.VerifyCodeVerifier(receivedCodeVerifier)
	fmt.Println("is the received code verifier valid?", isValid)
}
```

Okay, so that was a bit easy... But, what can I configure?!?

```go
package main

import (
	"fmt"

	"github.com/matthewhartstonge/pkce"
)

func main() {
	// Generate a ... proof key!
	key, err := pkce.New(
		// pkce.WithCodeVerifierLength enables increasing entropy for 
		// super-duper securities!
		pkce.WithCodeVerifierLength(9001),

		// pkce.WithChallengeMethod enables setting the PKCE mode, which is 
		// really code name for setting the method to "plain" for, you know, if
		// you've got a non-compliant OAuth PKCE accepting server that may 
		// require backwards compatibility. #SnarkIntended
		pkce.WithChallengeMethod(pkce.Plain),

		// pkce.WithCodeVerifier enables BYO code verifier.
		//
		// ... I hope you use a secure implementation ...
		//
		// This is mainly useful if you like the struct style of encapsulation, 
		// or if loading the verifier from a datastore.
		//
		// Using this option will disable code verifier generation, therefore 
		// `pkce.WithCodeVerifierLength` will be redundant if specified.
		pkce.WithCodeVerifier([]byte("#YOLO")),
	)
	if err != nil {
		// hah, yeah, there's gonna be an error or two...
		panic(err)
	}

	// ... otherwise, it's business as usual ...
```

### Functional

For those that like functions, you can fight against your own "to err == programmer"

```go
package main

import (
	"fmt"

	"github.com/matthewhartstonge/pkce"
)

func main() {
	// Generate a secure code verifier!
	codeVerifier, err := pkce.GenerateCodeVerifier(50)
	if err != nil {
		panic(err)
	}

	// OR, lawd forbid, you can generate and send in your own code verifier...
	//    ... don't do this ...

	// Then we can generate a code challenge based on the incoming code 
	// challenge method
	codeChallenge, err := pkce.GenerateCodeChallenge(pkce.S256, codeVerifier)
	if err != nil {
		panic(err)
	}

	fmt.Println("my manually generated code verifier is:", codeVerifier)
	fmt.Println("my manually generated code challenge is:", codeChallenge)

	// Finally - on the server-side side, we can verify the received code 
	// verifier:
	incomingCodeVerifier := "#yolo-cant-verify-me-mr-mcbaggins"
	isValid := pkce.VerifyCodeVerifier(pkce.S256, incomingCodeVerifier, codeChallenge)
	fmt.Println("is the received code verifier valid?", isValid)
}

```

## What is PKCE?

Great Question!

For more information on "Proof Key for Code Exchange (PKCE) by OAuth Public Clients" (or for some light bedtime reading)
check out the following links:

* [RFC 7636](https://datatracker.ietf.org/doc/html/rfc7636)
* [Auth0 - How PKCE Works](https://auth0.com/docs/get-started/authentication-and-authorization-flow/authorization-code-flow-with-proof-key-for-code-exchange-pkce#how-it-works)
* [OAuth - "Protecting Apps with PKCE"](https://www.oauth.com/oauth2-servers/pkce/)


CHANGELOG.md
# Changelog
All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/),
and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).

## [Unreleased]
## [v0.1.2] - 2022-01-27
### Added
- :white_check_mark: pkce: adds tests.

### Changed
- :recycle: options: orders options alphabetically.

### Fixed
- :bug: options: fixes challenge method being able to be set with invalid values.
- :bug: pkce: stops invalid methods from being able to be set.
- :bug: pkce: stops the code verifier length being overwritten if code verifier is set.
- :bug: pkce: ensures getCodeVerifier checks against nilness and emptiness.

## [v0.1.1] - 2022-01-25
### Added
- :memo: README: adds code verifier verification examples.
- :page_facing_up: LICENSE: adds MIT license. Fixes #4.
- :construction_worker: ci/cd: enables github actions. Fixes #3.
- :memo: README: adds go reference, go report and github action build badges.

### Fixed
- :memo: README: fixes a couple of spelling misteaks.
- :rotating_light: pkce: fixes whitespace issues (wsl).
- :rotating_light: validation: simplifies if-return (revive).
- :rotating_light: errors: reduces line lengths (lll).

## [v0.1.0] - 2022-01-25
### Added
- Generation of code verifier and code challenge.
- Code verifier and code challenge spec-compliant validation.
- Verification for an incoming code verifier.
- URL parameter key constants.

[Unreleased]: https://github.com/matthewhartstonge/pkce/compare/v0.1.2...HEAD
[v0.1.2]: https://github.com/matthewhartstonge/pkce/compare/v0.1.1...v0.1.2
[v0.1.1]: https://github.com/matthewhartstonge/pkce/compare/v0.1.0...v0.1.1
[v0.1.0]: https://github.com/matthewhartstonge/pkce/releases/tag/v0.1.0


options_test.go
package pkce

import (
	"reflect"
	"testing"
)

func TestWithChallengeMethod(t *testing.T) {
	tests := setChallengeMethodTests()

	for _, tt := range tests {
		t.Run(tt.name, func(t *testing.T) {
			opt := WithChallengeMethod(tt.method)

			err := opt(tt.gotKey)
			if (err != nil) != tt.shouldErr {
				t.Errorf("WithChallengeMethod() should error\ngot:  %v, want: %v\n", err, tt.shouldErr)
			}

			if tt.shouldErr {
				if tt.wantErr != err {
					t.Errorf("WithChallengeMethod() error type not expected\ngot:  %v, want: %v\n", err, tt.wantErr)
				}
			} else {
				if !reflect.DeepEqual(tt.gotKey, tt.wantKey) {
					t.Errorf("WithChallengeMethod() key\ngot: %v\nwant  %v\n", tt.gotKey, tt.wantKey)
				}
			}
		})
	}
}

func TestWithCodeVerifier(t *testing.T) {
	tests := setCodeVerifierTests()

	for _, tt := range tests {
		t.Run(tt.name, func(t *testing.T) {
			opt := WithCodeVerifier(tt.codeVerifier)

			err := opt(tt.gotKey)
			if (err != nil) != tt.shouldErr {
				t.Errorf("WithCodeVerifier() should error\ngot:  %v, want: %v\n", err, tt.shouldErr)
			}

			if tt.shouldErr {
				if tt.wantErr != err {
					t.Errorf("WithCodeVerifier() error type not expected\ngot:  %v, want: %v\n", err, tt.wantErr)
				}
			} else {
				if !reflect.DeepEqual(tt.gotKey, tt.wantKey) {
					t.Errorf("WithCodeVerifier() key\ngot: %v\nwant  %v\n", tt.gotKey, tt.wantKey)
				}
			}
		})
	}
}

func TestWithCodeVerifierLength(t *testing.T) {
	tests := setCodeVerifierLengthTests()

	for _, tt := range tests {
		t.Run(tt.name, func(t *testing.T) {
			opt := WithCodeVerifierLength(tt.n)

			err := opt(tt.gotKey)
			if (err != nil) != tt.shouldErr {
				t.Errorf("WithCodeVerifierLength() should error\ngot:  %v, want: %v\n", err, tt.shouldErr)
			}

			if tt.shouldErr {
				if tt.wantErr != err {
					t.Errorf("WithCodeVerifierLength() error type not expected\ngot:  %v, want: %v\n", err, tt.wantErr)
				}
			} else {
				if !reflect.DeepEqual(tt.gotKey, tt.wantKey) {
					t.Errorf("WithCodeVerifierLength() key\ngot: %v\nwant  %v\n", tt.gotKey, tt.wantKey)
				}
			}
		})
	}
}


validation_test.go
package pkce

import (
	"crypto/rand"
	"fmt"
	"strings"
	"testing"
)

func Test_validateCodeVerifier(t *testing.T) {
	type args struct {
		verifier []byte
	}

	tests := []struct {
		name      string
		args      args
		shouldErr bool
		wantErr   error
	}{
		{
			name: "should error if verifier is nil",
			args: args{
				verifier: nil,
			},
			shouldErr: true,
			wantErr:   ErrVerifierLength,
		},
		{
			name: "should error if verifier is empty",
			args: args{
				verifier: []byte{},
			},
			shouldErr: true,
			wantErr:   ErrVerifierLength,
		},
		{
			name: "should error if verifier is too short",
			args: args{
				verifier: []byte(strings.Repeat("a", verifierMinLen-1)),
			},
			shouldErr: true,
			wantErr:   ErrVerifierLength,
		},
		{
			name: "should error if verifier is too long",
			args: args{
				verifier: []byte(strings.Repeat("a", verifierMaxLen+1)),
			},
			shouldErr: true,
			wantErr:   ErrVerifierLength,
		},
		{
			name: "should error with invalid characters (ascii)",
			args: args{
				verifier: []byte(strings.Repeat("a", verifierMinLen-1) + "!"),
			},
			shouldErr: true,
			wantErr:   ErrVerifierCharacters,
		},
		{
			name: "should error with invalid characters (utf-8)",
			args: args{
				verifier: []byte("aaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaa💩"),
			},
			shouldErr: true,
			wantErr:   ErrVerifierCharacters,
		},
		{
			name: "should error with invalid characters (random byte array)",
			args: args{
				verifier: randomBytes(t, 100),
			},
			shouldErr: true,
			wantErr:   ErrVerifierCharacters,
		},
		{
			name: "should pass with a short verifier",
			args: args{
				verifier: []byte(strings.Repeat("a", verifierMinLen)),
			},
			shouldErr: false,
		},
		{
			name: "should pass with long verifier",
			args: args{
				verifier: []byte(strings.Repeat("a", verifierMaxLen)),
			},
			shouldErr: false,
		},
	}

	for _, tt := range tests {
		t.Run(tt.name, func(t *testing.T) {
			err := validateCodeVerifier(tt.args.verifier)
			if (err != nil) != tt.shouldErr {
				t.Errorf("validateCodeVerifier() should have error\ngot:  %v\nwant: %v\n", err, tt.shouldErr)
			}
			if (err != nil) && tt.wantErr != err {
				t.Errorf("validateCodeVerifier() expected error\ngot:  %v\nwant: %v\n", err, tt.wantErr)
			}
		})
	}
}

func Test_validateVerifierLen(t *testing.T) {
	type args struct {
		n int
	}

	tests := []struct {
		name      string
		args      args
		shouldErr bool
		wantErr   error
	}{
		{
			name: "should error if value is negative",
			args: args{
				n: -1,
			},
			shouldErr: true,
			wantErr:   ErrVerifierLength,
		},
		{
			name: fmt.Sprintf("should error if length is smaller than min length (%d)", verifierMinLen),
			args: args{
				n: verifierMinLen - 1,
			},
			shouldErr: true,
			wantErr:   ErrVerifierLength,
		},
		{
			name: fmt.Sprintf("should consider min length + 1 (%d) valid", verifierMinLen+1),
			args: args{
				n: verifierMinLen + 1,
			},
			shouldErr: false,
		},
		{
			name: fmt.Sprintf("should consider max length - 1 (%d) valid", verifierMaxLen-1),
			args: args{
				n: verifierMinLen + 1,
			},
			shouldErr: false,
		},
		{
			name: fmt.Sprintf("should error if length is greater than max length (%d)", verifierMaxLen),
			args: args{
				n: verifierMaxLen + 1,
			},
			shouldErr: true,
			wantErr:   ErrVerifierLength,
		},
	}

	for _, tt := range tests {
		t.Run(tt.name, func(t *testing.T) {
			err := validateVerifierLen(tt.args.n)
			if (err != nil) != tt.shouldErr {
				t.Errorf("validateVerifierLen() should have error\ngot:  %v\nwant: %v\n", err, tt.shouldErr)
			}
			if (err != nil) && tt.wantErr != err {
				t.Errorf("validateVerifierLen() expected error\ngot:  %v\nwant: %v\n", err, tt.wantErr)
			}
		})
	}
}

func Test_validateCodeVerifierCharacters(t *testing.T) {
	type args struct {
		chars []byte
	}

	tests := []struct {
		name      string
		args      args
		shouldErr bool
		wantErr   error
	}{
		{
			name: "should error with invalid characters (ascii)",
			args: args{
				chars: []byte("abc123!abc123"),
			},
			shouldErr: true,
			wantErr:   ErrVerifierCharacters,
		},
		{
			name: "should error with invalid characters (utf-8)",
			args: args{
				chars: []byte("XÆA-Xii"),
			},
			shouldErr: true,
			wantErr:   ErrVerifierCharacters,
		},
		{
			name: "should error with invalid characters (random byte array)",
			args: args{
				chars: randomBytes(t, 100),
			},
			shouldErr: true,
			wantErr:   ErrVerifierCharacters,
		},
		{
			name: "should pass with valid characters",
			args: args{
				chars: []byte(unreserved),
			},
			shouldErr: false,
		},
	}

	for _, tt := range tests {
		t.Run(tt.name, func(t *testing.T) {
			err := validateCodeVerifierCharacters(tt.args.chars)
			if (err != nil) != tt.shouldErr {
				t.Errorf("validateCodeVerifierCharacters() should have error\ngot:  %v\nwant: %v\n", err, tt.shouldErr)
			}
			if (err != nil) && tt.wantErr != err {
				t.Errorf("validateCodeVerifierCharacters() expected error\ngot:  %v\nwant: %v\n", err, tt.wantErr)
			}
		})
	}
}

func Test_validVerifierChar(t *testing.T) {
	type args struct {
		c byte
	}

	type validVerifierCharTest struct {
		name string
		args args
		want bool
	}

	tests := []validVerifierCharTest{}
	for _, char := range []byte(unreserved) {
		tests = append(tests, validVerifierCharTest{
			name: fmt.Sprintf("should consider ASCII reserved character '%c' valid", char),
			args: args{
				c: char,
			},
			want: true,
		})
	}

	for _, char := range []byte("`!@#$%^&*()+={}[]|\\/:;<>,?") {
		tests = append(tests, validVerifierCharTest{
			name: fmt.Sprintf("should consider ASCII character '%c' invalid", char),
			args: args{
				c: char,
			},
			want: false,
		})
	}

	// given UTF-8 chunks overflows a byte, we can test against a possible,
	// albeit extremely weird, vector.
	for _, char := range []byte("💩🐶あ") {
		tests = append(tests, validVerifierCharTest{
			name: "should consider UTF-8 characters invalid",
			args: args{
				c: char,
			},
			want: false,
		})
	}

	for _, tt := range tests {
		t.Run(tt.name, func(t *testing.T) {
			if got := validVerifierChar(tt.args.c); got != tt.want {
				t.Errorf("validVerifierChar() = %v, want %v", got, tt.want)
			}
		})
	}
}

func randomBytes(t *testing.T, n int) (randomBytes []byte) {
	randomBytes = make([]byte, n)
	_, err := rand.Read(randomBytes)
	if err != nil {
		t.Error("unable to generate random byte stream:", err)
	}

	return
}


errors.go
package pkce

import (
	"errors"
	"fmt"
)

var (
	// ErrMethodDowngrade enforces compliance with RFC 7636, 7.2.
	//
	// Clients MUST NOT downgrade to "plain" after trying the "S256" method.
	// Servers that support PKCE are required to support "S256", and servers
	// that do not support PKCE will simply ignore the unknown
	// "code_verifier".  Because of this, an error when "S256" is presented
	// can only mean that the server is faulty or that a MITM attacker is
	// trying a downgrade attack.
	ErrMethodDowngrade = errors.New("clients must not downgrade to 'plain' after trying the 'S256' method")

	// ErrMethodNotSupported enforces the use of compliant transform methods
	ErrMethodNotSupported = errors.New("clients must use either 'plain' or 'S256' as a transform method")

	// ErrVerifierCharacters enforces character compliance with the unreserved
	// character set as specified in RFC 7636, 4.1.
	ErrVerifierCharacters = fmt.Errorf(
		"code verifier must only contain unreserved characters from the set: {'%s'}",
		unreserved,
	)

	// ErrVerifierLength enforces compliance with the minimum and maximum
	// lengths as specified in RFC 7636, 4.1.
	ErrVerifierLength = fmt.Errorf(
		"code verifier must be between %d and %d characters long",
		verifierMinLen,
		verifierMaxLen,
	)
)


options.go
package pkce

// Option enables variadic PKCE Key options to be configured.
type Option func(*Key) error

// WithChallengeMethod enables specifying the challenge transformation method.
// Should only be used to downgrade to plain if required.
func WithChallengeMethod(method Method) Option {
	return func(key *Key) (err error) {
		switch method {
		case Plain, S256:
			key.challengeMethod = method

		default:
			return ErrMethodNotSupported
		}

		return nil
	}
}

// WithCodeVerifier enables supplying your own code verifier. Disables code
// verifier generation.
func WithCodeVerifier(codeVerifier []byte) Option {
	return func(key *Key) (err error) {
		// validate incoming code verifier
		err = key.setCodeVerifier(codeVerifier)

		return
	}
}

// WithCodeVerifierLength enables specifying the length of the code verifier
// to be generated.
func WithCodeVerifierLength(n int) Option {
	return func(key *Key) (err error) {
		err = key.setCodeVerifierLength(n)

		return
	}
}


validation.go
package pkce

// validateCodeVerifier ensures that the provided code verifier is specification
// compliant.
func validateCodeVerifier(verifier []byte) error {
	if err := validateVerifierLen(len(verifier)); err != nil {
		return err
	}

	return validateCodeVerifierCharacters(verifier)
}

// validateVerifierLen ensures the length of the code verifier is within the
// bounds of the specification's declared lengths.
func validateVerifierLen(n int) error {
	if n < verifierMinLen || n > verifierMaxLen {
		return ErrVerifierLength
	}

	return nil
}

// validateCodeVerifier ensures all characters provided are in the set of
// unreserved characters.
func validateCodeVerifierCharacters(chars []byte) error {
	for _, char := range chars {
		if !validVerifierChar(char) {
			return ErrVerifierCharacters
		}
	}

	return nil
}

// validVerifierChar ensures that any bytes provided are specifically from the
// unreserved character list.
func validVerifierChar(c byte) bool {
	if 'a' <= c && c <= 'z' || 'A' <= c && c <= 'Z' || '0' <= c && c <= '9' {
		return true
	}

	switch c {
	case '-', '.', '_', '~':
		return true
	}

	return false
}


pkce.go
// Package pkce implements proof key generation defined by RFC 7636 to enable
// generation and validation of code verifiers and code challenges.
//
// "Proof Key for Code Exchange" (PKCE, pronounced "pixy") was created as a
// technique to mitigate against the authorization code interception attack.
//
// For a detailed specification of PKCE (RFC 7636) see [1].
//
// Terminology:
//   1. code verifier
//     A cryptographically random string that is used to correlate the
//     authorization request to the token request.
//
//   2. code challenge
//     A challenge derived from the code verifier that is sent in the
//     authorization request, to be verified against later.
//
//   3. code challenge method
//     A method that was used to derive code challenge.
//
//   4. Base64url Encoding
//     Base64 encoding using the URL- and filename-safe character set
//     defined in Section 5 of [RFC4648], with all trailing '='
//     characters omitted (as permitted by Section 3.2 of [RFC4648]) and
//     without the inclusion of any line breaks, whitespace, or other
//     additional characters.  (See Appendix A for notes on implementing
//     base64url encoding without padding.)
//
// [1] https://datatracker.ietf.org/doc/html/rfc7636
package pkce

import (
	"crypto/rand"
	"crypto/sha256"
	"encoding/base64"
	"math/big"
)

// Method specifies the code challenge transformation method that was used to
// derive the code challenge.
type Method string

// String implements Stringer.
func (m Method) String() string {
	return string(m)
}

const (
	// Plain method specifies that the code challenge has had no transformation
	// performed on the code verifier.
	//
	// code_challenge = code_verifier
	//
	// The plain transformation is for compatibility with existing
	// deployments and for constrained environments that can't use the S256
	// transformation.
	//
	Plain Method = "plain"

	// S256 method specifies that the code challenge has been transformed by
	// being hashed by SHA-256 then base64url-encoded.
	//
	// code_challenge = BASE64URL-ENCODE(SHA256(ASCII(code_verifier)))
	//
	// If the client is capable of using "S256", it MUST use "S256", as
	// "S256" is Mandatory To Implement (MTI) on the server.  Clients are
	// permitted to use "plain" only if they cannot support "S256" for some
	// technical reason and know via out-of-band configuration that the
	// server supports "plain".
	S256 Method = "S256"
)

const (
	// ABNF for "code_verifier"
	// ALPHA = %x41-5A / %x61-7A
	alpha = "ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz"
	// DIGIT = %x30-39
	digit = "0123456789"
	// unreserved = ALPHA / DIGIT / "-" / "." / "_" / "~"
	unreserved = alpha + digit + "-._~"
)

const (
	// RFC 7636, 4.1
	verifierMinLen = 43
	verifierMaxLen = 128
)

// New returns a Proof Key
func New(opts ...Option) (key *Key, err error) {
	key = &Key{
		challengeMethod: S256,
		codeVerifierLen: verifierMinLen,
	}

	for _, opt := range opts {
		if err = opt(key); err != nil {
			return
		}
	}

	return
}

// GenerateCodeVerifier generates an RFC7636 compliant, cryptographically secure
// code verifier.
func GenerateCodeVerifier(n int) (string, error) {
	if err := validateVerifierLen(n); err != nil {
		return "", err
	}

	return string(generateCodeVerifier(n)), nil
}

// GenerateCodeChallenge takes a code verifier and method to generate a code
// challenge.
func GenerateCodeChallenge(method Method, codeVerifier string) (out string, err error) {
	in := []byte(codeVerifier)
	if err = validateCodeVerifier(in); err != nil {
		return
	}

	return generateCodeChallenge(method, in), nil
}

// VerifyCodeVerifier enables servers to verify the received code verifier.
func VerifyCodeVerifier(method Method, codeVerifier string, codeChallenge string) bool {
	// RFC 7636, 4.6.
	//
	// the server verifies it by calculating the code challenge from the
	// received "code_verifier" and comparing it with the previously associated
	// "code_challenge", after first transforming it according to the
	// "code_challenge_method" method specified by the client.
	switch method {
	case Plain:
		// If the "code_challenge_method" from Section 4.3 was "plain", they are
		// compared directly, i.e.:
		return codeVerifier == codeChallenge

	case S256:
		// If the "code_challenge_method" from Section 4.3 was "S256", the
		// received "code_verifier" is hashed by SHA-256, base64url-encoded, and
		// then compared to the "code_challenge", i.e.:
		codeVerifierChallenge, err := GenerateCodeChallenge(method, codeVerifier)
		if err != nil {
			return false
		}

		return codeVerifierChallenge == codeChallenge

	default:
		return false
	}
}

// Key provides the proof key for secure code exchange.
type Key struct {
	// challengeMethod determines the code challenge transform method to use.
	challengeMethod Method
	// codeVerifierLen provides the length of the code verifier to generate, if
	// a code verifier is not supplied on key generation.
	codeVerifierLen int
	// codeVerifier provides the code verifier data.
	codeVerifier []byte
}

// SetChallengeMethod enables upgrading code challenge generation method.
func (k *Key) SetChallengeMethod(method Method) error {
	switch method {
	case Plain, S256:
		if k.challengeMethod == S256 && method == Plain {
			return ErrMethodDowngrade
		}

		k.challengeMethod = method

	default:
		return ErrMethodNotSupported
	}

	return nil
}

// ChallengeMethod returns the configured key's method for generating a code
// challenge.
func (k *Key) ChallengeMethod() Method {
	return k.challengeMethod
}

// setCodeVerifierLength sets the length of the code verifier to be generated.
//
// If a code verifier is supplied, this setting will be ignored in favour of
// using the supplied verifier.
func (k *Key) setCodeVerifierLength(n int) error {
	if len(k.codeVerifier) > 0 {
		// Don't overwrite the set length.
		return nil
	}

	if err := validateVerifierLen(n); err != nil {
		return err
	}

	k.codeVerifierLen = n

	return nil
}

// setCodeVerifier enables setting a new code verifier.
func (k *Key) setCodeVerifier(verifier []byte) (err error) {
	if err = validateCodeVerifier(verifier); err != nil {
		return
	}

	k.codeVerifier = verifier
	k.codeVerifierLen = len(verifier)

	return
}

// CodeVerifier returns the code verifier.
func (k *Key) CodeVerifier() string {
	return string(k.getCodeVerifier())
}

// getCodeVerifier returns a code verifier. If one has not been set, it will
// generate one based on the configured verifier length.
func (k *Key) getCodeVerifier() []byte {
	if len(k.codeVerifier) == 0 {
		k.codeVerifier = generateCodeVerifier(k.codeVerifierLen)
	}

	return k.codeVerifier
}

// CodeChallenge returns the challenge for the configured code verifier.
// Will generate a verifier if nil.
func (k *Key) CodeChallenge() string {
	return generateCodeChallenge(k.ChallengeMethod(), k.getCodeVerifier())
}

// VerifyCodeVerifier provides a convenience function, for if you've loaded the
// code verifier into the key. If not, this won't really be useful to use...
func (k *Key) VerifyCodeVerifier(codeVerifier string) bool {
	return VerifyCodeVerifier(k.ChallengeMethod(), codeVerifier, k.CodeChallenge())
}

// generateCodeVerifier performs the computations required to generate a
// cryptographically random, specification compliant code verifier.
func generateCodeVerifier(n int) (out []byte) {
	unreservedLen := big.NewInt(int64(len(unreserved)))

	out = make([]byte, n)
	for i := range out {
		// ensure we use non-deterministic random ints.
		j, _ := rand.Int(rand.Reader, unreservedLen)
		out[i] = unreserved[j.Int64()]
	}

	return out
}

// generateCodeChallenge performs the transform required by the specified
// method.
func generateCodeChallenge(method Method, codeVerifier []byte) (out string) {
	if method == Plain {
		return string(codeVerifier)
	}

	s256 := sha256.New()
	s256.Write(codeVerifier)

	return base64.RawURLEncoding.EncodeToString(s256.Sum(nil))
}
